src/portable/synopsys/dwc2/dwc2_esp32.h

```diff
@@ -82,6 +82,22 @@ static const dwc2_controller_t _dwc2_controller[] =
 #endif
 };
 
+static ep_avaliable_count_t ep_avaliable_count[] =
+{
+#ifdef DWC2_FS_PERIPH_BASE
+// ESP32Sx has 6 endpoints, from which only 5 can be confiugred as IN
+#if TU_CHECK_MCU(OPT_MCU_ESP32S2, OPT_MCU_ESP32S3)
+  { .out_ep = DWC2_FS_EP_MAX, .in_ep = DWC2_FS_EP_MAX - 1},
```

Comment:
Shouldn't it be  `OTG_NUM_IN_EPS - 1` here? According to the databook:

- `OTG_NUM_EPS` is defined as "the number of Device mode endpoints in addition to endpoint 0, which is
always present". Meaning we have
  - EP0 IN, EP0 OUT (two endpoints) always present
  - Another `OTG_NUM_EPS` endpoints available
- `OTG_NUM_IN_EPS` is defined as "the maximum number of Device mode IN endpoints active at any time including endpoint 0". Meaning that...
  - Total number of endpoints available is `OTG_NUM_EPS + 2`
  - Of those total endpoints, only `OTG_NUM_IN_EPS` can be made IN at any one point in time, one of them is already EP0 IN.

Reply:
I was following endpoint definition from esp32s3 datasheet.

Device Mode Features

- Endpoint number 0 always present (bi-directional, consisting of EP0 IN and EP0 OUT)
- Six additional endpoints (endpoint numbers 1 to 6), configurable as IN or OUT
- Maximum of five IN endpoints concurrently active at any time (including EP0 IN)

Now, I am taking a better look. From what it is written in the datasheet:

-  on the S3, we have 6 configurable EPs and one Bi-directional EP0
- from which 5 could be IN EPs, together with the always present EP0
- There are 5 TX FIFOs on the S3, from which one is ocuppied by EP0
- thus we only have 4 IN EPs avaliable to be opened by the TinyUSB

Is that correct? Or did I misunderstood?

Reply:
Generally speaking, Darian is right. 

Although, here we open the EPs which have not 0 number. (We handle opening EP0 in dcd_dwc2, `bus_reset()` and we do not need to relate to maximum EPs number there and increase the amount of using EPs).

So we don't need to take the EP0 into consideration here, when we opening EPs.